File: 92. Reverse Linked List II.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
struct ListNode* reverseBetween(struct ListNode* head, int m, int n) {
    int i;
    struct ListNode *prev, *cur, *tmp, *p;

    if (m == n) {
        return head;
    } else {
        i = 1;
        cur = head;
        prev = NULL;

        /* Find m-th node. */
        while (i < m && cur) {
            prev = cur;
            cur = cur->next;
            i++;
        }
        p = prev;
        /* reverse linked list while finding n-th node. */
        while (i <= n && cur) {
            tmp = cur->next;
            cur->next = prev;
            prev = cur;
            cur = tmp;
            i++;
        }

        /* If m = 1 */
        if (p == NULL) {
            head->next = cur;
            head = prev;
        } else {
            /* If m > 1. */
            p->next->next = cur;
            p->next = prev;
        }
        return head;
    }
}

File: Python3/92. Reverse Linked List II.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
        def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
            # Locate m-th node
            dummy, dummy.next = ListNode(-1), head
            prev, cur = dummy, head
            for _ in range(m - 1):
                prev, cur = cur, cur.next

            # Reverse nodes from m-th to n-th
            tmp, prev, cur = prev, cur, cur.next
            for _ in range(n - m):
                cur.next, cur, prev = prev, cur.next, cur
            tmp.next.next, tmp.next = cur, prev
            return dummy.next
